net: tcp: Cancel fin_timer in FIN_WAIT2 instead FIN_WAIT1

mniestroj opened the pull request:
According to RFC 793 we should wait for FIN in FIN-WAIT-1 and
FIN-WAIT-2 states. Receiving ACK in FIN-WAIT-1 just moves us to
FIN-WAIT-2 state.

Right now TCP connection is never closed if FIN is not received
in FIN-WAIT-2 state. Fix that by keeping fin_timer active in
FIN-WAIT-2 state, but canceling it just after FIN is received.

Fixes: 124c06702726 ("net: tcp: Cancel the fin_timer on FIN message
  in FIN_WAIT1 state")
Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

jukkar reviewed (changes requested):
Good catch. After looking the TCP state diagram, there seems to be some issues still with the code.
If we receive FIN,ACK in FIN_WAIT_1 we should go to TIME_WAIT directly, in this case we should cancel the fin_timer already in FIN_WAIT_1.
Your change to cancel the timer in FIN_WAIT_2 is ok, we just need some adjustment for the FIN_WAIT_1 case.

jukkar commented:
Note that PR #6718 moved TCP code from net_context.c to tcp.c so a rebase is needed.

mniestroj commented:
Thanks for review. I will prepare a rebase next week. Do you want FIN handling in FIN_WAIT_1 state in this PR? If so, then do you have some suggestions how to test FIN,ACK in FIN_WAIT_1 case? This is my first experience with TCP implementation, I just tried to fix the case that was easily reproducible in my usage scenario.

jukkar commented:
> If so, then do you have some suggestions how to test FIN,ACK in FIN_WAIT_1 case?

I was thinking something like this below, please modify if needed. Note that this is not really tested, I just quickly outlined something. Please test this so we can see if it works as expected.

```
Author: Jukka Rissanen <jukka.rissanen@linux.intel.com>
Date:   Tue Apr 3 11:05:18 2018 +0300

net: tcp: Properly handle FIN timer in FIN_WAIT_1 and FIN_WAIT_2
    
According to RFC 793 we should wait for FIN in FIN_WAIT_1 and
FIN_WAIT_2 states. Receiving plain ACK in FIN_WAIT_1 just moves
us to FIN_WAIT_2 state.
    
Right now TCP connection is never closed if FIN is not received
in FIN_WAIT_2 state. Fix that by keeping fin_timer active in
FIN_WAIT_2 state, but canceling it just after FIN is received.
    
If we receive FIN,ACK in FIN_WAIT_1 state, then the fin_timer is
cancelled already in FIN_WAIT_1 as we move directly to TIME_WAIT
state.
    
Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>
Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

diff --git a/subsys/net/ip/tcp.c b/subsys/net/ip/tcp.c
index 79b4becf53..fc58d91c13 100644
--- a/subsys/net/ip/tcp.c
+++ b/subsys/net/ip/tcp.c
@@ -2004,12 +2004,22 @@ NET_CONN_CB(tcp_established)
                 * list, e.g., an ack of FIN is received.
                 */
 
-               if (net_tcp_get_state(context->tcp)
-                          == NET_TCP_FIN_WAIT_1) {
-                       /* Received FIN on FIN_WAIT1, so cancel the timer */
-                       k_delayed_work_cancel(&context->tcp->fin_timer);
-                       /* Active close: step to FIN_WAIT_2 */
-                       net_tcp_change_state(context->tcp, NET_TCP_FIN_WAIT_2);
+               if (net_tcp_get_state(context->tcp) == NET_TCP_FIN_WAIT_1) {
+                       if (tcp_flags & NET_TCP_FIN) {
+                               /* If we receive FIN + ACK then go directly to
+                                * TIME_WAIT state. In this case we cancel the
+                                * FIN timer. If ACK was not set, then the timer
+                                * will be cancelled in FIN_WAIT_2 state.
+                                */
+                               k_delayed_work_cancel(&context->tcp->fin_timer);
+
+                               net_tcp_change_state(context->tcp,
+                                                    NET_TCP_TIME_WAIT);
+                       } else {
+                               /* Active close: step to FIN_WAIT_2 */
+                               net_tcp_change_state(context->tcp,
+                                                    NET_TCP_FIN_WAIT_2);
+                       }
                } else if (net_tcp_get_state(context->tcp)
                           == NET_TCP_LAST_ACK) {
                        /* Passive close: step to CLOSED */
@@ -2034,6 +2044,8 @@ NET_CONN_CB(tcp_established)
                                              ACK_TIMEOUT);
                                              ACK_TIMEOUT);
                } else if (net_tcp_get_state(context->tcp)
                           == NET_TCP_FIN_WAIT_2) {
+                       /* Received FIN on FIN_WAIT1, so cancel the timer */
+                       k_delayed_work_cancel(&context->tcp->fin_timer);
                        /* Active close: step to TIME_WAIT */
                        net_tcp_change_state(context->tcp, NET_TCP_TIME_WAIT);
                }

```

mniestroj commented:
I don't think we gain anything by handling "FIN + ACK" case and moving directly from FIN_WAIT_1 to TIME_WAIT state. Instead we add more code, which makes it harder in future to understand and improve. The final output of both approaches (mine and yours) is the same, but in mine approach you have just small overhead of switching to TIME_WAIT via FIN_WAIT_2 state, instead swtiching directly. I don't think it is worth optimizing.

The case that is not handled right now is receiving FIN in FIN_WAIT_1 state. But for now I do not know how to test such case, so I do not have any propositions yet how to implement that. I also think that we can do this in future independently of current issue.

jukkar commented:
> I don't think we gain anything by handling "FIN + ACK" case and moving directly from FIN_WAIT_1 to TIME_WAIT state. Instead we add more code, which makes it harder in future to understand and improve. The final output of both approaches (mine and yours) is the same, but in mine approach you have just small overhead of switching to TIME_WAIT via FIN_WAIT_2 state, instead swtiching directly. I don't think it is worth optimizing.

I do not agree with this. The issue is not the few extra lines of code that we might add, also the code complexity is imho not increased too much. The issue is that if we do not handle the FIN+ACK, then there will be a timeout because the peer needs to send the final ACK again to us.

mniestroj commented:
> The issue is that if we do not handle the FIN+ACK, then there will be a timeout because the peer needs to send the final ACK again to us.
What timeout are you writing about?

As I understand in case of "FIN+ACK", after switching from FIN_WAIT_1 to FIN_WAIT_2 we will immediately disable fin_timer and switch to TIME_WAIT state. This is exactly the same as handling "FIN+ACK" directly in your code. Am I wrong? What am I missing?

jukkar commented:
> What timeout are you writing about?

I am talking about the connection timeout from the peer. The peer is expecting that we send ACK, so it might send FIN+ACK again in this case.

pfalcon commented:
@mniestroj :

> I will prepare a rebase next week.

May I join @jukkar's suggestion to make such a rebase, it's easier to invoke suggestions like:

> I also think that we can do this in future independently of current issue.

If you have a mergeable patch ;-).

pfalcon commented:
@jukkar 

> If we receive FIN,ACK in FIN_WAIT_1 we should go to TIME_WAIT directly, in this case we should cancel the fin_timer already in FIN_WAIT_1.

Please remember that RFC793 stipulates that every TCP segment sent with TCP session established should have ACK sent. During session teardown, the session is apparently still established, so *any* FIN would also have ACK set. Please also mind that "ACK" flag means nothing more than "acknowledge sequence number field is valid in this segment". So, the reading of state diagram for "recv: FIN, ACK" shouldn't be as simple as "segment has both FIN and ACK flags set" (it always has, per above), but: "segment has FIN, and an ACK seqno for the FIN we sent previously".

That's a more complicated check, and I would agree with @mniestroj - this PR as it stands is already an improvement and can go as is (after rebasing), and further improvements may need future work (and additional research/testing).

mniestroj committed:
net: tcp: Cancel fin_timer in FIN_WAIT_2 instead FIN_WAIT_1

According to RFC 793 we should wait for FIN in FIN_WAIT_1 and
FIN_WAIT_2 states. Receiving ACK in FIN_WAIT_1 just moves us to
FIN_WAIT_2 state.

Right now TCP connection is never closed if FIN is not received
in FIN_WAIT_2 state. Fix that by keeping fin_timer active in
FIN_WAIT_2 state, but canceling it just after FIN is received.

Fixes: 124c06702726 ("net: tcp: Cancel the fin_timer on FIN message
  in FIN_WAIT1 state")
Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj commented:
@pfalcon Rebase done :)

jukkar commented:
> this PR as it stands is already an improvement

That is true, I am not arguing that. I just noticed during review that we are not handling this ACK+FIN combination properly. It would make lot of sense to fix that combination in this PR too (could be separate patch), otherwise this thing is easily forgotten.

But we can do this in smaller steps too. One thing I noticed in the original commit message is that it talks about "FIN-WAIT-1" instead of "FIN_WAIT_1", so please change the state names to have underscores in them as that is how the code is using them. Also in the change itself there is comment about "FIN_WAIT1", please change that to "FIN_WAIT_2".  Lets merge this after these changes and continue with the fin+ack case after that.
